### Class Assignment 2
#### Name: Anushka Sawant
#### Matriculation Number: 100006644

In [1]:
import spacy
from spacy import displacy
from IPython.display import HTML, display

#   Load the English language model
nlp = spacy.load("en_core_web_sm")

text = "I saw the man with a telescope."

doc = nlp(text)

# Print the dependency parse information for each token
for token in doc:
    print(f"{token.text:<12} {token.dep_:<12} {token.head.text}")

html = displacy.render(doc, style="dep", jupyter=False)

# Display the dependency parse visualization in Jupyter Notebook
display(HTML(html))

# Discuss with your partner:
# 1. What is the ROOT?
# 2. What does "with a telescope" modify?
# 3. Does the sentence have more than one meaning?
# 4. Did spaCy choose the interpretation you expected?
# 5. Rewrite the sentence to remove ambiguity.

I            nsubj        saw
saw          ROOT         saw
the          det          man
man          dobj         saw
with         prep         saw
a            det          telescope
telescope    pobj         with
.            punct        saw


## INtERPRETATION
#### 1. What is the ROOT?
- Ans: Saw
#### 2. What does "with a telescope" modify?
- Ans: "with a telescope" can modify either "saw" (indicating how the action of seeing was performed) or "man" (indicating that the man has a telescope).
#### 3. Does the sentence have more than one meaning?
- Ans: Yes, this is syntactic ambiguity. The sentence can be interpreted in two main ways:
Meaning A: I used a telescope as a tool to see the man. (The telescope modifies saw).
Meaning B: I saw a man, and that man was carrying or possessing a telescope. (The telescope modifies man).
#### 4. Did spaCy choose the interpretation you expected?
- Ans: spacy chose Meaning A
#### 5. Rewrite the sentence to remove ambiguity.
- Ans: I saw the man who was holding a telescope.

In [2]:
import spacy

nlp = spacy.load("en_core_web_sm")

text = """
Microsoft announced a new AI research initiative
in Berlin on 15 June 2026.

The project involves collaboration with Google,
SAP and SRH University Heidelberg.

Satya Nadella will present the initiative
at a conference in Germany.
"""

doc = nlp(text)

for ent in doc.ents:
    print(
        ent.text,
        ent.label_,
        spacy.explain(ent.label_)
    )

html = displacy.render(doc, style="ent", jupyter=False)
display(HTML(html))
 
#1. Which entities were detected?
#2. Which labels were assigned?
#3. Which entity types occur most often?
#4. Were any entities missed?

Microsoft ORG Companies, agencies, institutions, etc.
AI ORG Companies, agencies, institutions, etc.
Berlin GPE Countries, cities, states
15 June 2026 DATE Absolute or relative dates or periods
Google ORG Companies, agencies, institutions, etc.
SAP ORG Companies, agencies, institutions, etc.
Satya Nadella PERSON People, including fictional
Germany GPE Countries, cities, states


## INTERPRETATION
#### 1. Which entities were detected?
- Ans:-Microsoft, Berlin, 15 June 2026, Google, SAP, SRH University Heidelberg, Satya Nadella, Germany
 
#### 2. Which labels were assigned?
- Ans:-ORG, GPE, DATE, ORG, ORG, ORG, PERSON, GPE
 
#### 3. Which entity types occur most often?
- Ans:-ORG (Organization)
 
#### 4. Were any entities missed?
- Ans:-Yes, 'SRH University Heidelberg' was missing

In [3]:
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

documents = [
    "AI is transforming healthcare",
    "Machine learning improves healthcare",
    "Football is a popular sport"
]

vectorizer = TfidfVectorizer()

tfidf_matrix = vectorizer.fit_transform(documents)

print("Vocabulary:")
print(vectorizer.get_feature_names_out())

print("\nTF-IDF Matrix:")
print(tfidf_matrix.toarray())

similarity_matrix = cosine_similarity(tfidf_matrix)

print("\nCosine Similarity Matrix:")
print(similarity_matrix)

print("Doc1 vs Doc2:", round(similarity_matrix[0][1], 3))
print("Doc1 vs Doc3:", round(similarity_matrix[0][2], 3))
print("Doc2 vs Doc3:", round(similarity_matrix[1][2], 3))

# 1. Which pair has the highest similarity?
# 2. Is this result reasonable?
# 3. Why does TF-IDF treat some words as more important?
# 4. Does cosine understand meaning?
# 5. Would it know that "AI" and "machine learning" are related?

Vocabulary:
['ai' 'football' 'healthcare' 'improves' 'is' 'learning' 'machine'
 'popular' 'sport' 'transforming']

TF-IDF Matrix:
[[0.5628291  0.         0.42804604 0.         0.42804604 0.
  0.         0.         0.         0.5628291 ]
 [0.         0.         0.40204024 0.52863461 0.         0.52863461
  0.52863461 0.         0.         0.        ]
 [0.         0.52863461 0.         0.         0.40204024 0.
  0.         0.52863461 0.52863461 0.        ]]

Cosine Similarity Matrix:
[[1.         0.17209173 0.17209173]
 [0.17209173 1.         0.        ]
 [0.17209173 0.         1.        ]]
Doc1 vs Doc2: 0.172
Doc1 vs Doc3: 0.172
Doc2 vs Doc3: 0.0


## INTERPRETATION
#### 1. Which pair has the highest similarity?
- Ans: There is a tie for the highest similarity between Doc1 vs Doc2 (0.172) and Doc1 vs Doc3 (0.172).

#### 2. Is this result reasonable?
- Ans: Yes, mathematically it is reasonable based on TF-IDF. Doc1 seems to share exactly one word with Doc2 and exactly one word with Doc3, resulting in identical overlap scores.

#### 3. Why does TF-IDF treat some words as more important?
- Ans: TF-IDF lowers the weight of words that appear frequently across all documents like "is" or "the" and increases the weight of words that appear frequently in a specific document but rarely across the whole dataset like "healthcare" or "football".

#### 4. Does cosine understand meaning?
- Ans: No. Cosine similarity calculated on a TF-IDF matrix only understands exact word overlap, it deos not understand the semantic meaning.

#### 5. Would it know that "AI" and "machine learning" are related?
- Ans: No. In a standard TF-IDF model, "AI" and "machine learning" are treated as completely independent features